narwhals/_duckdb/dataframe.py

```diff
@@ -161,13 +186,124 @@ def _change_version(self: Self, version: Version) -> Self:
 
     def _from_native_frame(self: Self, df: Any) -> Self:
         return self.__class__(
-            df, version=self._version, backend_version=self._backend_version
+            df, backend_version=self._backend_version, version=self._version
+        )
+
+    def group_by(self: Self, *keys: str, drop_null_keys: bool) -> DuckDBGroupBy:
+        from narwhals._duckdb.group_by import DuckDBGroupBy
+
+        if drop_null_keys:
+            msg = "todo"
+            raise NotImplementedError(msg)
+
+        return DuckDBGroupBy(
+            compliant_frame=self, keys=list(keys), drop_null_keys=drop_null_keys
         )
 
+    def rename(self: Self, mapping: dict[str, str]) -> Self:
+        df = self._native_frame
+        selection = [
+            f"{col} as {mapping[col]}" if col in mapping else col for col in df.columns
+        ]
+        return self._from_native_frame(df.select(", ".join(selection)))
+
+    def join(
+        self: Self,
+        other: Self,
+        *,
+        how: Literal["left", "inner", "outer", "cross", "anti", "semi"] = "inner",
+        left_on: str | list[str] | None,
+        right_on: str | list[str] | None,
+        suffix: str,
+    ) -> Self:
+        if isinstance(left_on, str):
+            left_on = [left_on]
+        if isinstance(right_on, str):
+            right_on = [right_on]
+
+        if how not in ("inner", "left"):
+            msg = "Only inner and left join is implemented for DuckDB"
+            raise NotImplementedError(msg)
+        assert left_on is not None  # noqa: S101
+        assert right_on is not None  # noqa: S101
+        conditions = []
+        lhs = []
+        for left, right in zip(left_on, right_on):
+            conditions.append(f"lhs.{left} = rhs.{right}")
+            lhs.append(left)
```

Comment:
As `left_on` is a list of strings:

```suggestion
        lhs = left_on.copy()
        conditions = [f"lhs.{left} = rhs.{right}" for left, right in zip(left_on, right_on)]
```